docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,278 @@
+
+These attributes, however are not the only mechanisms available to the custom marshaller author, the source generator insepct the marshaller itself for various other indications that inform how marshalling should occur.
+
+Complete details on the design can be found in the [dotnet/runtime][design_libraryimport] repository.
+
+### Source generator Analyzer and Fixer
+
+Along with the source generator itself, an Analyzer and Fixer are both provided. The Analyzer and Fixer are enabled and available by default since .NET 7 RC1. The Analyzer is designed to help guide developers into using the source generator properly and the Fixer provides automated conversions from many `DllImport` patterns into the appropriate `LibraryImport` signature.
+
+## Introducing the native library
+
+Using the `LibraryImport` source generator would mean consuming a native, or unmanaged, library. A native library might be a shared library (that is, `.dll`, `.so`, or `dylib`) that directly calls an operating system API, not exposed through .NET, or one that is heavily optimized in an unmanaged language that a .NET developer wants to consume. For this tutorial we are going to build our own shared library that exposes a C style API surface. The below represents a user defined type and two APIs that we would like to consume from C#. These two APIs represent the "in" mode, but there are additional modes to explore in the sample.
+
+```cpp
+struct error_data
+{
+    int code;
+    bool is_fatal_error;
+    char32_t* message;    /* UTF-32 encoded string */
+};
+
+extern "C" DLL_EXPORT void STDMETHODCALLTYPE PrintString(char32_t* chars);
+
+extern "C" DLL_EXPORT void STDMETHODCALLTYPE PrintErrorData(error_data data);
+```
+
+The above contains the two types of interest, `char32_t*` and `error_data`. The latter represents a string that is encoded in UTF-32, not an string encoding .NET historically marshals, and a user defined type containing a 32-bit integer field, a C++ boolean field, and a UTF-32 encoded string field. Both of these types will require us to provide a way for the source generator to generate marshalling code.
+
+## Customize marshalling for a built-in type
+
+Let's look at the `char32_t*` type first, since marshalling this is required by the user-defined type. We should note that `char32_t*` represents the native side, but we also need representation in managed code. In .NET, there is only one "string" type, `string`. Therefore we will be marshalling a native UTF-32 encoded string to and from the `string` type in managed code. There are already several built-in marshallers for the `string` type that will marshal as UTF-8, UTF-16, ANSI, and even as the Windows `BSTR` type. However, there isn't one for marshalling as UTF-32. That is what we need to define.
```

Comment:
```suggestion
Consider the `char32_t*` type first, since marshalling this type is required by the user-defined type. `char32_t*` represents the native side, but you also need representation in managed code. In .NET, there's only one "string" type, `string`. Therefore, you'll be marshalling a native UTF-32 encoded string to and from the `string` type in managed code. There are already several built-in marshallers for the `string` type that marshal as UTF-8, UTF-16, ANSI, and even as the Windows `BSTR` type. However, there isn't one for marshalling as UTF-32. That's what you need to define.
```